apps/src/blockly/addons/cdoSpritePointer.js

```diff
@@ -0,0 +1,120 @@
+      // If this block has itself as a root and is in a flyout workspace,
+      // it is a in a mini toolbox. A block can't be moved from one flyout to
+      // another, so if it's in a flyout, it is in its original flyout.
+      if (blockWorkspace && blockWorkspace.isFlyout && block.imageSourceId) {
+        imageSourceBlock = mainWorkspace.getBlockById(block.imageSourceId);
+      }
+    }
+  }
+  if (imageSourceBlock) {
+    return getImageUrlFromImageSource(imageSourceBlock, pointerData.imageIndex);
+  } else {
+    // The block is probably disconnected from any root or toolbox. Reset to
+    // default text.
+    return '';
+  }
+}
+
+// Given a shadow block and an optional image source block id, update the image displayed
+// on the pointer block to match the image selected in the image source block. See getPointerBlockImageUrl
+// for details on how we find the image source block.
+export function updatePointerBlockImage(
+  block,
+  pointerMetadataMap,
+  imageSourceId
+) {
+  const url = getPointerBlockImageUrl(block, pointerMetadataMap, imageSourceId);
+  changePointerImage(url, block);
+}
+
+/**
+ * Get the image url from the image source block's inputList at the given index.
+ * We find the image url by looking at the connection to the input at the given index.
+ *
+ * @param {Block} imageSourceBlock Image source block to get image url from.
+ * @param {number} imageIndexOnSource index of the source block's input list to get the image url from. This index
+ *  should have a connection to a block of type gamelab_allSpritesWithAnimation, or we will return an empty string.
+ * @returns Image url or empty string if the image source block does not have an image at the given index.
+ */
+function getImageUrlFromImageSource(imageSourceBlock, imageIndexOnSource) {
+  const targetConnection =
+    imageSourceBlock.inputList[imageIndexOnSource]?.connection
+      ?.targetConnection;
+  // We only want to get the image from a connection block that is not an insertion marker.
+  // If the block is an insertion marker that means the block is being dragged
+  // and is not yet connected to the image source block.
+  if (
+    targetConnection &&
+    targetConnection.sourceBlock_ &&
+    targetConnection.sourceBlock_.type === 'gamelab_allSpritesWithAnimation' &&
+    !targetConnection.sourceBlock_.isInsertionMarker()
+  ) {
+    // Blocks of type gamelab_allSpritesWithAnimation have an input with one field (the costume
+    // picker dropdown).
+    return (
+      targetConnection.sourceBlock_.inputList[0].fieldRow[0].imageElement_.getAttribute(
+        'xlink:href'
+      ) || ''
+    );
+  } else {
+    return '';
+  }
+}
+
+// Set the given block's image input to the given imageUrl,
+// or reset the image input to the default long text if the imageUrl is empty.
+function changePointerImage(imageUrl, block) {
+  if (!imageUrl || imageUrl.length === 0) {
+    resetPointerImageToLongString(block);
```

Comment:
nit: Would it make more sense to call this `resetPointerImageToLongText` since you use a string as a placeholder for an empty image above?

Reply:
Nevermind, I see `longString` is a property. Disregard! (And I saw your comment clarifies.)